diff --git a/xla/service/cpu/cpu_instruction_fusion.h b/xla/service/cpu/cpu_instruction_fusion.h
index 1e090cb6..81b1dd9e 100644
--- a/xla/service/cpu/cpu_instruction_fusion.h
+++ b/xla/service/cpu/cpu_instruction_fusion.h
@@ -26,8 +26,13 @@ namespace cpu {
 
 class CpuInstructionFusion : public InstructionFusion {
  public:
+#if defined(ENABLE_ANNC) && defined(__aarch64__)
+  CpuInstructionFusion(bool may_duplicate)
+      : InstructionFusion(CpuInstructionFusion::IsExpensive, may_duplicate) {}
+#else
   CpuInstructionFusion()
       : InstructionFusion(CpuInstructionFusion::IsExpensive) {}
+#endif
   ~CpuInstructionFusion() override = default;
 
   using HloPassInterface::Run;
