diff --git a/third_party/xla/xla/service/cpu/cpu_runtime.h b/third_party/xla/xla/service/cpu/cpu_runtime.h
index e483b8c9..0d9afb06 100644
--- a/third_party/xla/xla/service/cpu/cpu_runtime.h
+++ b/third_party/xla/xla/service/cpu/cpu_runtime.h
@@ -85,6 +85,7 @@ extern const char* const kTracingStartSymbolName;
 extern const char* const kTracingEndSymbolName;
 extern const char* const kAllToAllSymbolName;
 extern const char* const kOneDnnMatMulSymbolName;
+extern const char* const kXnnPackSoftMaxNDSymbolName;
 
 // All symbol names for XLA CPU runtime functions need to start with this
 // prefix.
