diff -git a/xla/service/cpu/ir_emitter.h b/xla/service/cpu/ir_emitter.h
index c4534fa6..76bfce60 100644
--- a/xla/service/cpu/ir_emitter.h
+++ b/xla/service/cpu/ir_emitter.h
@@ -192,6 +192,9 @@ class IrEmitter : public DfsHloVisitorWithDefault,
   Status HandleTopK(HloInstruction* hlo);
   Status HandleAllReduceSingleReplica(HloInstruction* crs);
   Status HandleAllReduceMultipleReplica(HloInstruction* crs);
+#if defined(ENABLE_ANNC) && defined(__aarch64__)
+  Status HandleXnnPackSoftMax(HloInstruction* hlo);
+#endif
 #if defined(INTEL_MKL) && defined(ENABLE_ONEDNN_V3)
   Status HandleOneDnnMatMul(HloInstruction* hlo);
 #endif  // INTEL_MKL && ENABLE_ONEDNN_V3

