diff --git a/xla/BUILD b/xla/BUILD
index 67e522fa..1825ebf1 100644
--- a/xla/BUILD
+++ b/xla/BUILD
@@ -1052,6 +1052,18 @@ xla_cc_test(
         ],
 )
 
+alias(
+    name = "is_aarch64",
+    actual = "@local_xla//xla/service/cpu:is_aarch64",
+    visibility = ["//visibility:public"],
+)
+
+alias(
+    name = "aarch64_and_annc_disabled",
+    actual = "@local_xla//xla/service/cpu:aarch64_and_annc_disabled",
+    visibility = ["//visibility:public"],
+)
+
 cc_library(
     name = "debug_options_flags",
     srcs = [
@@ -1059,7 +1071,11 @@ cc_library(
         "debug_options_parsers.h",
     ],
     hdrs = ["debug_options_flags.h"],
-    copts = if_enable_acl(["-DXLA_CPU_USE_ACL=1"]),
+    copts = if_enable_acl(["-DXLA_CPU_USE_ACL=1"]) + select({
+        ":aarch64_and_annc_disabled": [],
+        ":is_aarch64": ["-DENABLE_ANNC"],
+        "//conditions:default": [],
+    }),
     visibility = ["//visibility:public"],
     deps =
         [

