--- ./tensorflow/compiler/xla/service/BUILD
+++ ./tensorflow/compiler/xla/service/BUILD
@@ -425,6 +425,7 @@
         "hlo_sharding.h",
         "hlo_sharding_metadata.h",
     ],
+    visibility = ["//visibility:public"],
     deps = [
         ":hlo_casting_utils",
         ":hlo_module_config",
@@ -519,6 +520,7 @@
 cc_library(
     name = "pattern_matcher",
     hdrs = ["pattern_matcher.h"],
+    visibility = ["//visibility:public"],
     deps = [
         ":hlo",
         ":hlo_casting_utils",
@@ -3365,6 +3367,7 @@
         "hlo_pass_fix.h",
         "hlo_pass_interface.h",
     ],
+    visibility = ["//visibility:public"],
     deps = [
         ":hlo",
         ":hlo_module_group",
@@ -4212,6 +4215,7 @@
     name = "hlo_parser",
     srcs = ["hlo_parser.cc"],
     hdrs = ["hlo_parser.h"],
+    visibility = ["//visibility:public"],
     deps = [
         ":hlo",
         ":hlo_casting_utils",
