diff --git a/tensorflow/core/grappler/optimizers/remapper.cc b/tensorflow/core/grappler/optimizers/remapper.cc
index 3c37150f..3fa204be 100644
--- a/tensorflow/core/grappler/optimizers/remapper.cc
+++ b/tensorflow/core/grappler/optimizers/remapper.cc
@@ -52,6 +52,9 @@ limitations under the License.
 #include "third_party/gpus/cudnn/cudnn.h"
 #endif  // GOOGLE_CUDA
 
+// Disable TensorFlow MatMul operator fusion
+#define DISABLE_TF_MATMUL_FUSION  1
+
 namespace tensorflow {
 namespace grappler {
 
@@ -3102,6 +3105,13 @@ Status AddFusedContractionNode(RemapperContext* ctx,
           << " bias_add=" << bias_add.name()
           << " contraction=" << contraction.name();
 
+#if defined(DISABLE_TF_MATMUL_FUSION)
+  if (IsMatMul(contraction)) {
+    VLOG(1) << " Disable MatMul operator fusion in TensorFlow. ";
+    return OkStatus();
+  }
+#endif
+
   NodeDef fused_op;
   fused_op.set_name(bias_add.name());
   fused_op.set_device(contraction.device());
@@ -3200,6 +3210,13 @@ Status AddFusedContractionNode(
   const NodeDef& bias_add = graph->node(matched.bias_add);
   const NodeDef& activation = graph->node(matched.activation);
 
+#if defined(DISABLE_TF_MATMUL_FUSION)
+  if (IsMatMul(contraction)) {
+    VLOG(1) << " Disable MatMul operator fusion in TensorFlow. ";
+    return OkStatus();
+  }
+#endif
+
   VLOG(2) << "Fuse " << contraction.op() << " with BiasAdd and "
           << activation.op() << ":"
           << " activation=" << activation.name()
@@ -3396,6 +3413,13 @@ Status AddFusedContractionNode(RemapperContext* ctx,
   DCHECK(IsConv2D(contraction) || IsMatMul(contraction) ||
          IsConv3D(contraction));
 
+#if defined(DISABLE_TF_MATMUL_FUSION)
+  if (IsMatMul(contraction)) {
+    VLOG(1) << " Disable MatMul operator fusion in TensorFlow. ";
+    return OkStatus();
+  }
+#endif
+
   NodeDef contraction_node;
   const NodeDef& add = graph->node(matched.add);
   contraction_node.set_name(add.name());
@@ -3600,6 +3624,13 @@ Status AddFusedMatMulBiasAddAndGelu(
   auto* matmul_node =
       ctx->graph_view.GetNode(matched_nodes_map.at("matmul"))->node();
 
+#if defined(DISABLE_TF_MATMUL_FUSION)
+  if (IsMatMul(*matmul_node)) {
+    VLOG(1) << " Disable MatMul operator fusion in TensorFlow. ";
+    return OkStatus();
+  }
+#endif
+
   NodeDef fused_node;
   // Fused node should have the name of terminal node of the fusion.
   fused_node.set_name(output_node->name());
diff --git a/third_party/xla/xla/service/cpu/BUILD b/third_party/xla/xla/service/cpu/BUILD
index 6e0ea613..28550bcd 100644
--- a/third_party/xla/xla/service/cpu/BUILD
+++ b/third_party/xla/xla/service/cpu/BUILD
@@ -203,6 +203,7 @@ cc_library(
     srcs = ["cpu_compiler.cc"],
     hdrs = ["cpu_compiler.h"],
     copts = tsl_copts(),
+    linkopts = ["-L/user/lib64", "-lannc"],
     visibility = ["//visibility:public"],
     deps = [
         ":buffer_info_util",
