diff --git a/tensorflow/core/BUILD b/tensorflow/core/BUILD
index 538574360ba..3864064d18c 100644
--- a/tensorflow/core/BUILD
+++ b/tensorflow/core/BUILD
@@ -629,6 +629,7 @@ cc_library(
         "//tensorflow/core/kernels/linalg",
         "//tensorflow/core/kernels/sparse:kernels",
         "//tensorflow/core/kernels/uniform_quant_ops:kernels",
+        "//tensorflow/core/kernels:embedding_fused_ops",
     ] + if_mkl([
         "//tensorflow/core/kernels/mkl:mkl_concat_op",
         "//tensorflow/core/kernels/mkl:mkl_dequantize_op",
@@ -907,6 +908,7 @@ filegroup(
         "candidate_sampling_ops_op_lib",
         "checkpoint_ops_op_lib",
         "clustering_ops_op_lib",
+        "embedding_fused_ops_op_lib",
         "collective_ops_op_lib",
         "control_flow_ops_op_lib",
         "count_ops_op_lib",
diff --git a/tensorflow/core/grappler/optimizers/BUILD b/tensorflow/core/grappler/optimizers/BUILD
index ecd559734ea..97a918ead6d 100644
--- a/tensorflow/core/grappler/optimizers/BUILD
+++ b/tensorflow/core/grappler/optimizers/BUILD
@@ -880,6 +880,18 @@ tf_cuda_cc_test(
     ],
 )
 
+alias(
+    name = "is_aarch64",
+    actual = "@local_xla//xla/service/cpu:is_aarch64",
+    visibility = ["//visibility:public"],
+)
+
+alias(
+    name = "aarch64_and_annc_disabled",
+    actual = "@local_xla//xla/service/cpu:aarch64_and_annc_disabled",
+    visibility = ["//visibility:public"],
+)
+
 tf_kernel_library(
     name = "remapper",
     srcs = ["remapper.cc"],
@@ -904,7 +916,12 @@ tf_kernel_library(
         "//tensorflow/core/grappler/utils:symbolic_shapes",
         "//tensorflow/core/grappler/utils:topological_sort",
         "@com_google_absl//absl/container:flat_hash_set",
-    ] + if_mkl(["//tensorflow/core/graph:mkl_graph_util"]),
+    ] + if_mkl(["//tensorflow/core/graph:mkl_graph_util"])
+    + select({
+        ":aarch64_and_annc_disabled": [],
+        ":is_aarch64": ["//tensorflow/core/grappler/optimizers/graph_optimizer:annc_graph_opt"],
+        "//conditions:default": [],
+    })
 )
 
 tf_cuda_cc_test(
diff --git a/tensorflow/core/grappler/optimizers/remapper.cc b/tensorflow/core/grappler/optimizers/remapper.cc
index 3c37150f496..2fbde836d1d 100644
--- a/tensorflow/core/grappler/optimizers/remapper.cc
+++ b/tensorflow/core/grappler/optimizers/remapper.cc
@@ -52,6 +52,7 @@ limitations under the License.
 #include "third_party/gpus/cudnn/cudnn.h"
 #endif  // GOOGLE_CUDA
 
+#include "tensorflow/core/grappler/optimizers/graph_optimizer/graph_opt.h"
 namespace tensorflow {
 namespace grappler {
 
@@ -4596,6 +4597,7 @@ bool RequiresInferredShapes(const RemapperContext& ctx, int node_index,
 Status Remapper::Optimize(Cluster* cluster, const GrapplerItem& item,
                           GraphDef* optimized_graph) {
   GrapplerItem mutable_item = item;
+  annc::run_graph_optimization(&mutable_item.graph);
   Status status;
   RemapperContext ctx(&mutable_item, &status, cpu_layout_conversion_,
                       xla_auto_clustering_on_);
diff --git a/tensorflow/core/kernels/BUILD b/tensorflow/core/kernels/BUILD
index 22617ef8ab5..f89a084f94c 100644
--- a/tensorflow/core/kernels/BUILD
+++ b/tensorflow/core/kernels/BUILD
@@ -3655,6 +3655,97 @@ tf_kernel_library(
     ]) + [":fft_impl"],
 )
 
+tf_kernel_library(
+    name = "embedding_fused_action_id_gather_op",
+    srcs = ["embedding_fused_action_id_gather.cc"],
+    deps = MATH_DEPS,
+)
+
+tf_kernel_library(
+    name = "embedding_fused_gather_op",
+    srcs = ["embedding_fused_gather.cc"],
+    deps = MATH_DEPS,
+)
+
+tf_kernel_library(
+    name = "embedding_fused_padding_op",
+    srcs = ["embedding_fused_padding.cc"],
+    deps = MATH_DEPS,
+)
+
+tf_kernel_library(
+    name = "embedding_fused_sparse_dynamic_stitch_op",
+    srcs = ["embedding_fused_sparse_dynamic_stitch.cc"],
+    deps = MATH_DEPS,
+)
+
+tf_kernel_library(
+    name = "embedding_fused_reshape_op",
+    srcs = ["embedding_fused_sparse_reshape.cc"],
+    deps = MATH_DEPS + [
+        ":reshape_util",
+    ],
+)
+
+tf_kernel_library(
+    name = "embedding_fused_sparse_segment_reduce_op",
+    srcs = ["embedding_fused_sparse_segment_reduce.cc"],
+    deps = MATH_DEPS,
+)
+
+tf_kernel_library(
+    name = "embedding_fused_sparse_segment_reduce_nonzero_op",
+    srcs = ["embedding_fused_sparse_segment_reduce_nonzero.cc"],
+    deps = MATH_DEPS + ["@com_google_absl//absl/container:flat_hash_map"],
+)
+
+tf_kernel_library(
+    name = "embedding_fused_sparse_select_op",
+    srcs = ["embedding_fused_sparse_select.cc"],
+    deps = MATH_DEPS,
+)
+
+tf_kernel_library(
+    name = "fused_batchmatmul_add_sigmoid_op",
+    srcs = ["fused_batchmatmul_add_sigmoid.cc"],
+    deps = MATH_DEPS,
+)

+cc_library(
+    name = "embedding_fused_ops",
+    deps = [
+        ":embedding_fused_action_id_gather_op",
+        ":embedding_fused_gather_op",
+        ":embedding_fused_padding_op",
+        ":embedding_fused_sparse_dynamic_stitch_op",
+        ":embedding_fused_reshape_op",
+        ":embedding_fused_sparse_segment_reduce_op",
+        ":embedding_fused_sparse_segment_reduce_nonzero_op",
+        ":embedding_fused_sparse_select_op",
+        ":fused_batchmatmul_add_sigmoid_op",
+    ],
+)
+
+tf_cc_test(
+    name = "embedding_fused_ops_test",
+    size = "small",
+    srcs = [
+        "embedding_fused_action_id_gather_test.cc",
+        "embedding_fused_sparse_dynamic_stitch_test.cc",
+        "embedding_fused_sparse_segment_reduce_test.cc",
+        "embedding_fused_sparse_segment_reduce_nonzero_test.cc",
+        "embedding_fused_padding_test.cc",
+        "embedding_fused_sparse_select_test.cc",
+        "embedding_fused_gather_test.cc",
+        "embedding_fused_sparse_reshape_test.cc",
+        "fused_batchmatmul_add_sigmoid_test.cc",
+    ],
+    deps = [
+        ":ops_testutil",
+        ":ops_util",
+        ":embedding_fused_ops",
+        "//tensorflow/core:core_cpu",
+        "//tensorflow/core:framework",
+        "//tensorflow/core:lib",
+        "//tensorflow/core:protos_all_cc",
+        "//tensorflow/core:test",
+        "//tensorflow/core:test_main",
+        "//tensorflow/core:testlib",
+    ],
+)
+
 tf_kernel_library(
     name = "reduction_ops",
     gpu_srcs = ["reduction_gpu_kernels.cu.h"],
diff --git a/tensorflow/core/ops/BUILD b/tensorflow/core/ops/BUILD
index 91d80b6c2b5..b00b2a5d027 100644
--- a/tensorflow/core/ops/BUILD
+++ b/tensorflow/core/ops/BUILD
@@ -55,6 +55,7 @@ tf_gen_op_libs(
         "decode_proto_ops",
         "encode_proto_ops",
         "experimental_dataset_ops",
+        "embedding_fused_ops",
         "filesystem_ops",
         "function_ops",
         "functional_ops",
@@ -323,6 +324,7 @@ cc_library(
         ":training_ops_op_lib",
         ":uniform_quant_ops_op_lib",
         ":word2vec_ops",
+        ":embedding_fused_ops_op_lib",
     ] + select({
         # Non-tpu platforms don't need tpu dependency.
         "//tensorflow:chromiumos": [],
diff --git a/tensorflow/python/BUILD b/tensorflow/python/BUILD
index 9810f8acd2e..3b0b23fec0a 100644
--- a/tensorflow/python/BUILD
+++ b/tensorflow/python/BUILD
@@ -269,6 +269,7 @@ py_strict_library(
         "//tensorflow/python/ops/signal:util_ops",
         "//tensorflow/python/ops/signal:window_ops",
         "//tensorflow/python/ops/structured:structured_ops",
+        "//tensorflow/python/ops:embedding_fused_ops_gen",
         "//tensorflow/python/platform:_pywrap_stacktrace_handler",
         "//tensorflow/python/platform:app",
         "//tensorflow/python/platform:client_testlib",
@@ -389,6 +390,7 @@ py_strict_library(
         "//tensorflow/python/ops:bitwise_ops",
         "//tensorflow/python/ops:boosted_trees_ops_gen",
         "//tensorflow/python/ops:clustering_ops_gen",
+        "//tensorflow/python/ops:embedding_fused_ops_gen",
         "//tensorflow/python/ops:composite_tensor_ops",
         "//tensorflow/python/ops:cond_v2",
         "//tensorflow/python/ops:cudnn_rnn_ops_gen",
diff --git a/tensorflow/python/ops/BUILD b/tensorflow/python/ops/BUILD
index b3a3d612a01..185da7da4a8 100644
--- a/tensorflow/python/ops/BUILD
+++ b/tensorflow/python/ops/BUILD
@@ -228,6 +228,14 @@ tf_gen_op_strict_wrapper_private_py(
     ],
 )

+tf_gen_op_strict_wrapper_private_py(
+    name = "embedding_fused_ops_gen",
+    visibility = ["//tensorflow:internal"],
+    deps = [
+        "//tensorflow/core:embedding_fused_ops_op_lib",
+    ],
+)
+
 tf_gen_op_strict_wrapper_private_py(
     name = "collective_ops_gen",
     visibility = ["//tensorflow:internal"],
diff --git a/third_party/xla/xla/service/cpu/BUILD b/third_party/xla/xla/service/cpu/BUILD
index 6e0ea613435..47c346b4e93 100644
--- a/third_party/xla/xla/service/cpu/BUILD
+++ b/third_party/xla/xla/service/cpu/BUILD
@@ -198,6 +198,23 @@ cc_library(
     ],
 )
 
+config_setting(
+    name = "disable_annc",
+    define_values = {"disable_annc": "false"},
+    visibility = ["//visibility:public"],
+)
+
+config_setting(
+    name = "is_aarch64",
+    constraint_values = ["@platforms//cpu:aarch64"],
+)
+
+config_setting(
+    name = "aarch64_and_annc_disabled",
+    constraint_values = ["@platforms//cpu:aarch64"],
+    define_values = {"disable_annc": "true"},
+)
+
 cc_library(
     name = "cpu_compiler_pure",
     srcs = ["cpu_compiler.cc"],